lib/land/trackers/api_tracker.rb

```diff
@@ -206,24 +205,82 @@ def user_agent
                      Land.config.blank_user_agent_string
 
         @user_agent = UserAgent[user_agent]
+        @user_agent.user_agent_type = UserAgentType['user']
 
         if Land.config.identify_crawlers && defined?(CrawlerDetect)
           crawler_detect = CrawlerDetect.new(user_agent)
-          user_agent_type = crawler_detect.is_crawler? ? 'crawl' : 'api'
-          @user_agent.update(user_agent_type: UserAgentType[user_agent_type])
+          @user_agent.user_agent_type = UserAgentType['crawl'] if crawler_detect.is_crawler?
         end
 
         browser = ::Browser.new(user_agent)
+        land_browser = Land::Browser[browser.name]
 
-        @user_agent.update(
-          browser: Browser[browser.name],
-          device: Device[browser.device.name],
-          platform: Platform[browser.platform.name],
-          browser_version: browser.version
-        )
+        @user_agent.browser = land_browser
+        @user_agent.device = Device[browser.device.name]
+        @user_agent.platform = Platform[browser.platform.name]
+        @user_agent.browser_version = browser.version
+        @user_agent.device_resolution_id = device_resolution&.id
+        @user_agent.browser_color_preference_id = browser_color_preference&.id
+
+        begin
+          @user_agent.save! if @user_agent.changed?
+          land_browser.save! if land_browser.changed?
+        rescue ActiveRecord::RecordNotUnique
+          retry
+        rescue ActiveRecord::RecordInvalid => e
+          retry if e.message == 'Validation failed: User agent has already been taken'
+
+          add_error_tag_to_land_span(e)
+        end
 
         @user_agent
       end
+
+      def device_resolution
+        device_width = params.dig('device_resolution', 'width')
+        device_height = params.dig('device_resolution', 'height')
+        device_orientation = params.dig('device_resolution', 'orientation')
+
+        return unless device_width && device_height && device_orientation
+
+        resolution = DeviceResolution.find_or_create_by(
+          device_resolution: "#{device_width}x#{device_height}",
+          width: device_width,
+          height: device_height,
+          orientation: device_orientation
+        )
+
+        resolution&.save! if resolution&.changed?
+        resolution
+      rescue ActiveRecord::RecordNotUnique
+        retry
```

Comment:
- the `find_or_create_by` is already saving, so the `resolution&.save! if resolution&.changed?` is redundant
- UserAgent should accept nested parameters for `device_resolution`, making the rescue here redundant as well, so changing to `find_or_initialize_by` and allowing the `user_agent.save` to save this record, and then that rescue handle the race condition

```suggestion
        DeviceResolution.find_or_initialize_by(
          device_resolution: "#{device_width}x#{device_height}",
          width: device_width,
          height: device_height,
          orientation: device_orientation
        )
```